src/gpu/generic/sycl/ref_convolution.hpp

```diff
@@ -0,0 +1,383 @@
+/*******************************************************************************
+* Copyright 2024 Intel Corporation
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#ifndef GPU_SYCL_REF_CONVOLUTION_HPP
+#define GPU_SYCL_REF_CONVOLUTION_HPP
+
+#include "gpu/generic/sycl/sycl_gpu_primitive.hpp"
+#include "gpu/generic/sycl/sycl_io_helper.hpp"
+#include "gpu/generic/sycl/sycl_post_ops.hpp"
+#include "gpu/generic/sycl/sycl_primitive_conf.hpp"
+#include "gpu/generic/sycl/sycl_q10n.hpp"
+#include "gpu/gpu_convolution_pd.hpp"
+#include "xpu/sycl/types.hpp"
+
+namespace dnnl {
+namespace impl {
+namespace gpu {
+namespace generic {
+namespace sycl {
+
+struct ref_convolution_fwd_t : public gpu::generic::sycl::primitive_t {
+    using gpu::generic::sycl::primitive_t::primitive_t;
+
+    struct pd_t : public convolution_fwd_pd_t {
+        using convolution_fwd_pd_t::convolution_fwd_pd_t;
+
+        DECLARE_COMMON_PD_T("dpcpp:ref:any", ref_convolution_fwd_t);
+
+        status_t init(impl::engine_t *engine) {
+            using namespace data_type;
+            using sm = primitive_attr_t::skip_mask_t;
+
+            const memory_desc_wrapper data_d(src_md());
+            const memory_desc_wrapper weights_d(weights_md());
+            const memory_desc_wrapper dst_d(dst_md());
+
+            const bool ok = is_fwd() && check_work_amount(weights_d)
+                    && set_default_formats()
+                    && attr_.set_default_formats(dst_md()) == status::success
+                    && check_data_types(data_d, weights_d, dst_d)
+                    && check_formats(data_d, weights_d, dst_d)
+                    && attr()->has_default_values(sm::scales_runtime
+                            | sm::zero_points_runtime | sm::post_ops
+                            | sm::sum_dt)
+                    && IMPLICATION(!attr()->scales_.has_default_values(),
+                            check_scales_mask())
+                    && post_ops_ok();
+            if (!ok) return status::unimplemented;
+
+            return init_conf();
+        }
+
+        sycl_convolution_conf_t conf_;
+
+    private:
+        status_t init_conf();
+
+        bool set_default_formats() {
```

Comment:
It seems the private member funcs are the same, if you do not foresee them changing it would DRY-up the code if they get moved to protected space of a base class?

Reply:
I decided against common base class, as these classes already have bases and multiple inheritance tends to be frowned upon. I instead used common free helper functions.